fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2024 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from abc import ABC, abstractmethod
+from typing import Callable
+import rclpy
+import rclpy.node as Node
+import rmf_adapter.easy_full_control as rmf_easy
+from fleet_adapter_mir.mir_api import MirAPI
+
+
+class MirAction(ABC):
+    def __init__(
+            self,
+            node,
+            name,
+            mir_api: MirAPI,
+            update_handle,
+            fleet_config,
+            action_config: dict | None,
+    ):
+        self.node = node
+        self.name = name
+        self.api = mir_api
+        self.update_handle = update_handle
+        self.fleet_config = fleet_config
+        self.action_config = action_config
+        self.actions = self.action_config.get('actions')
+
+        missions_json = self.action_config.get('missions_json')
+        if missions_json:
+            with open(missions_json, 'r') as g:
+                action_missions = json.load(g)
+
+            # Check if these missions are already created on the robot
+            missions_created = True
+            for mission_name in action_missions.keys():
+                if mission_name not in self.api.known_missions:
+                    missions_created = False
+                    break
+            if missions_created:
+                return
+
+            # Create these missions on the robot
+            self.api.create_missions(action_missions)
+            # Update mission actions stored in MirAPI
+            for mission, mission_data in self.api.known_missions.items():
+                self.api.mission_actions[mission] = \
+                    self.api.missions_mission_id_actions_get(
+                        mission_data['guid'])
+
+    # This will be called whenever an action has begun
+    @abstractmethod
+    def perform_action(self,
```

Comment:
It's a bit clunky that the same action instance could have `perform_action` called on it multiple times. I don't think that [affordance](https://hackernoon.com/affordance-in-software-design-12cc0d9d2721) matches the intention of the API.

It might be better to merge this function with `ActionFactory.make_action`. Perhaps we could have something like the following:

```
class ActionFactory(ABC):
    ...
    def perform_action(
        self,
        category: str,
        description: dict,
        context: ActionContext
    ) -> MirAction:
```
I think `execution` could go inside of `ActionContext`.

Then `ActionFactory` would be responsible for looking at the category and description and then initializing the action accordingly. As soon as the `MirAction` is created, it begins performing.